spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -1236,6 +1236,9 @@ public void run() {
 				catch (Exception e) {
 					handleConsumerException(e);
 				}
+				finally {
+					afterRecordsProcessed();
```

Comment:
I see now why no records in the "after" contract. 
Probably that `pollAndInvoke()` may return that `ConsumerRecords`?
Yes, it might be a `null` since we may fail doing all of that hard logic, but still might be better to propagate such a context down to this "after" contract invocation.

Reply:
I guess the same comment about redundancy applies here too - the batch interceptor gets the `ConsumerRecord` ini success or failure and the record interceptor gets each record individually.

Reply:
Perhaps the issue is naming of `afterRecordsProcessed()` method. Perhaps we should remove `records` from it and call it `afterPollProcessed()` or similar?

Reply:
If we go with a new `BeforeAfterPollProcessor` contract, then just `afterPoll()` would be enough. 
Something in parallel with that `beforePoll()`.

Reply:
👍 

I am planning on building the next milestone (2.8.0-M3) later today; if would be great if we can get this into that ready for the next Boot milestone next week.

It's not really due until Monday, but I will be on PTO next week.

Reply:
I don't think `afterPoll()` sounds right; it implies it's done right after. `afterPollProcessed()` sounds good.

Reply:
or `afterPollResultsProcessed()`.

Reply:
OK. Then I'm good with that `afterPollProcessed()`.

Reply:
`AfterRollbackProcessor` and few other classes have `clearThreadState()` method already. Maybe instead of `afterPoll()` it should be `clearThreadState()` hear as well?

Reply:
I could live with `clearThreadState()`.

I am not sure `ConsumerPollProcessor` conveys the right meaning.

(I agree that naming is hard 😄 ).

Still thinking - let's stick with `BeforeAfterPollProcessor.clearThreadState()` for now.

It's only a milestone, so we are at liberty to change it if we can come up with something better.

...time passes...

How about something more general...

```java
public ThreadScopeProcessor {

    void setupThreadState();

   void clearThreadState();

}
```

??

Reply:
Unfortunately it seems that we won't be able to do `AfterRollbackProcessor extends ThreadScopeProcessor` because `AfterRollbackProcessor` lacks the setup side. I like the generic `ThreadScopeProcessor` as an idea but it would have to be applied in more place to be valid. If it's just this one place I like other names that mention "poll" better.

I finished refactor the code and used `BeforeAfterPollProcessor`. It's getting late here in London so I will leave the renaming of the classes to you. I'm happy with either option.

Reply:
I think we could make the `setUpThreadState` a default method; the DARP and stateful error handlers set up thread state as a side-effect of their normal processing.

Reply:
Thanks for the work - I can take it from here.

Reply:
Few minor fixed to compiler warnings, but now I'm done. Have fun, and thanks for helping me with the PR.